_pages/documentation/general_docs/developement/coding_style.md

```diff
@@ -0,0 +1,378 @@
+
+Always include Python.h first if you need that header. This is mandated by the integration guide. The next header file should be your main header file (e.g., for foo.cc you'd include foo.hh first). Having this header first ensures that it is independent and can be included in other places without missing dependencies.
+```c++
+// Include Python.h first if you need it.
+#include <Python.h>
+
+// Include your main header file before any other non-Python headers (i.e., the one with the same name as your cc source file)
+#include "main_header.hh"
+
+// C includes in sorted order
+#include <fcntl.h>
+#include <sys/time.h>
+
+// C++ includes
+#include <cerrno>
+#include <cstdio>
+#include <string>
+#include <vector>
+
+// Shared headers living in include/. These are used both in the simulator and utilities such as the m5 tool.
+#include <gem5/asm/generic/m5ops.h>
+
+// M5 includes
+#include "base/misc.hh"
+#include "cpu/base.hh"
+#include "params/BaseCPU.hh"
+#include "sim/system.hh"
+```
+
+## File structure and modularity
+
+Source files (.cc files) should never contain extern declarations; instead, include the header file associated with the .cc file in which the object is defined. This header file should contain extern declarations for all objects exported from that .cc file. This header should also be included in the defining .cc file. The key here is that we have a single external declaration in the .hh file that the compiler will automatically check for consistency with the .cc file. (This isn't as important in C++ as it was in C, since linker name mangling will now catch these errors, but it's still a good idea.)
+
+When sufficient (i.e., when declaring only pointers or references to a class), header files should use forward class declarations instead of including full header files.
+
+Header files should never contain using namespace declarations at the top level. This forces all the names in that namespace into the global namespace of any source file including that header file, which basically completely defeats the point of using namespaces. It is OK to use using namespace declarations at the top level of a source (.cc) file since the effect is entirely local to that .cc file. It's also OK to use them in _impl.hh files, since for practical purposes these are source (not header) files despite their extension. 
+
+## Documenting the code
+
+Each file/class/member should be documented using doxygen style comments.Doxygen allows users to quickly create documentation for our code by extracting the relavent information from the code and comments. It is able to document all the code structures including classes, namespaces, files, members, defines, etc. Most of these are quite simple to document, you only need to place a special documentation block before the declaration. The Doxygen documentation within gem5 is processed every night and the following web pages are generated: [Doxygen](http://www.gem5.org/docs/html/index.html) 
```

Comment:
The doxygen link is wrong now. Basically, we can't link to *anything* on gem5.org since that's going to go away. @BobbyRBruce, what's the link we should be using?